src/garage/torch/algos/vpg.py

```diff
@@ -0,0 +1,286 @@
+        self.env_spec = env_spec
+        self.policy = policy
+        self.baseline = baseline
+        self.max_path_length = max_path_length
+        self.policy_lr = policy_lr
+        self.discount = discount
+        self.gae_lambda = gae_lambda
+        self.center_adv = center_adv
+        self.positive_adv = positive_adv
+        self.max_kl_step = max_kl_step
+        self.policy_ent_coeff = policy_ent_coeff
+        self.use_softplus_entropy = use_softplus_entropy
+        self.stop_entropy_gradient = stop_entropy_gradient
+        self.entropy_method = entropy_method
+        self.eps = 1e-8
+
+        self.maximum_entropy = (entropy_method == 'max')
+        self.entropy_regularzied = (entropy_method == 'regularized')
+        #     Entropy regularization, max entropy rl
+
+        self._check_entropy_configuration(entropy_method, center_adv,
+                                          stop_entropy_gradient,
+                                          policy_ent_coeff)
+
+        self.reward_mean = collections.deque(maxlen=100)
+
+        self.optimizer = optimizer(
+            self.policy.get_module().parameters(), lr=self.policy_lr)
+
+    def _check_entropy_configuration(self, entropy_method, center_adv,
+                                     stop_entropy_gradient, policy_ent_coeff):
+        if entropy_method not in ('max', 'regularized', 'no_entropy'):
+            raise ValueError('Invalid entropy_method')
+
+        if entropy_method == 'max':
+            if center_adv:
+                raise ValueError('center_adv should be False when '
+                                 'entropy_method is max')
+            if not stop_entropy_gradient:
+                raise ValueError('stop_gradient should be True when '
+                                 'entropy_method is max')
+        if entropy_method == 'no_entropy':
+            if policy_ent_coeff != 0.0:
+                raise ValueError('policy_ent_coeff should be zero '
+                                 'when there is no entropy method')
+
+    def train_once(self, itr, paths):
+        """Perform one step of policy optimization."""
+        self._optimize_policy(itr, paths)
+        self._optimize_baseline(itr, paths)
+
+        return np.mean([sum(path['rewards']) for path in paths])
+
+    def _optimize_policy(self, itr, paths):
+        loss = [self._get_policy_loss(path) for path in paths]
+
+        self.optimizer.zero_grad()
+        torch.cat(loss).mean().backward()
+        self.optimizer.step()
+
+    def _get_policy_loss(self, path):
+        rewards = torch.Tensor(path['rewards'])
+        obs = torch.Tensor(path['observations'])
+
+        policy_entropy = self._get_policy_entropy(obs)
+
+        if self.maximum_entropy:
+            rewards = [
+                reward + self.policy_ent_coeff * policy_entropy
+                for reward in rewards
+            ]
+
+        baselines = self._get_baselines(path)
+        path['baselines'] = baselines.detach().numpy()
```

Comment:
I think we should always make `baseline.predict` detach so we don't need to detach here.